agent/runtime: Add seccomp feature

ManaSugi opened the pull request:
agent/runtime: add seccomp support
The agent supports seccomp feature based on the [OCI runtime specification](https://github.com/opencontainers/runtime-spec/blob/master/config-linux.md).
The feature can be enabled in the agent's `Makefile`.
If the agent is compiled with `SECCOMP=yes`, the feature is included in the build target and enabled.

In addition, it is better to remove `supports_seccomp_flag` in `agent.proto` because the agent started seccomp support from this commit. 
The kata container is ready for the seccomp support, so the agent and runtime do not need the flag anymore.

Fixes: #1476

Signed-off-by: Manabu Sugimoto <Manabu.Sugimoto@sony.com>

ManaSugi commented:
## Experiments of the seccomp feature

Apply seccomp profile to kata container using [crictl](https://github.com/kubernetes-sigs/cri-tools/blob/master/docs/crictl.md)  
Restricted system calls in seccomp profile
- `getuid`
- `chdir`
- `dup2`
  - The `dup2` syscall will fail only the new file descriptor is zero (stdin).

For more details of the seccomp profile, please sse the `seccomp.json` in the following config files.

### Conduct experiment using `crictl`

```
$ sudo crictl run -r kata container.json sandbox.json
---
97b17cdf8fe4e3093f8a695dd9b4d1a60167326edc021018de6516e3805a830b

$ sudo crictl attach --stdin --tty 97b17cdf8fe4e3093f8a695dd9b4d1a60167326edc021018de6516e3805a830b
# id
uid=4294967295 gid=0(root) euid=0(root) groups=0(root)

# cd
/usr/bin/sh: 2: cd: can't cd to /root

# ls -l | grep home
ls: write error: Broken pipe
```

Each command inside the container failed due to the seccomp profile.

### config files
<details><summary>seccomp.json</summary>

```json
{
	"defaultAction": "SCMP_ACT_ALLOW",
	"architectures": [
		"SCMP_ARCH_X86_64",
		"SCMP_ARCH_X86"
	],
	"flags": [ "SECCOMP_FILTER_FLAG_TSYNC" ],
	"syscalls": [
		{
			"names": [
				"getuid",
				"chdir"
			],
			"action": "SCMP_ACT_ERRNO"
		},
		{
			"names": [
				"dup2"
			],
			"action": "SCMP_ACT_ERRNO",
			"args": [
				{
					"index": 1,
					"value": 0,
					"op": "SCMP_CMP_EQ"
				}
			]
		}
	]
}
```

</details>

<details><summary>sandbox.json</summary>

```json
{
    "metadata": {
        "name": "ubuntu-pod",
        "uid": "ubuntu-pod",
        "namespace": "test.kata"
    },
    "hostname": "ubuntu_host",
    "log_directory": "",
    "dns_config": {
    },
    "port_mappings": [],
    "resources": {
    },
    "labels": {
    },
    "log_path": "/var/run/kata.log",
    "linux": {
    }
}
```

</details>

<details><summary>container.json</summary>

```json
{
    "metadata": {
        "name": "ubuntu",
        "namespace": "test.kata"
    },
    "image": {
        "image": "docker.io/ubuntu"
    },
    "command": [
        "/usr/bin/sh"
    ],
    "working_dir": "/",
    "envs": [
        {
            "key": "PATH",
            "value": "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
        },
        {
            "key": "TERM",
            "value": "xterm"
        }
    ],
    "stdin": true,
    "stdin_once": true,
    "tty": true,
    "log_path": "/var/run/kata.log",
    "linux": {
        "security_context": {
            "seccomp_profile_path": "localhost/<path of seccomp.json>/seccomp.json"
        }
    }
}
```

</details>

### Evironment information
<details><summary>Environment information</summary>

```
$ uname -a
Linux #### 5.8.0-50-generic #56~20.04.1-Ubuntu SMP Mon Apr 12 21:46:35 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

$ containerd --version
containerd github.com/containerd/containerd v1.4.0-beta.1-55-gc7518074 c75180740937d4b2d44b9c1edc1c27b208e66e32

$ crictl --version
crictl version v1.21.0

libseccomp version
2.5.1
```

</details>

## How to enable seccomp feature
To enable seccomp feature, we should take the following 3 steps.

1. Enable runtime seccomp support
  - Set `disable_guest_seccomp` as `false` in configuration.toml
  - e.g. `$ sudo sed -i '/^disable_guest_seccomp/ s/true/false/' /etc/kata-containers/configuration.toml`
1. Setup for libseccomp
   1. Installing the [libseccomp library](https://github.com/seccomp/libseccomp)
      - e.g. On debian-based linux  
      `$ sudo apt install libseccomp-dev`
   1. Set the environment variables for [libseccomp crate](https://crates.io/crates/libseccomp)
      - We should use the crate with statically linked the libseccomp library because the agent runs inside the VM.  It is possible to use the crate with dynamically linked the library, but we must set the correct path of `libseccomp.so` inside the kata image. That is not best solution because we may need to set `rpath`.  
      `$ export LIBSECCOMP_LINK_TYPE=static`  
      `$ export LIBSECCOMP_LIB_PATH="the path of libseccomp.a"`
1. Enable agent seccomp support
  - To include the seccomp feature in the build target, the feature should be enabled in the agent's Makefile at the [step](https://github.com/kata-containers/kata-containers/blob/main/docs/Developer-Guide.md#build-a-custom-kata-agent---optional).  
  `$ cd $GOPATH/src/github.com/kata-containers/kata-containers/src/agent` 
  `$ make SECCOMP=yes`
  - In addition, we have to set SECCOMP=yes when we create a local rootfs at the [step](https://github.com/kata-containers/kata-containers/blob/main/docs/Developer-Guide.md#create-a-local-rootfs-for-initrd-image).  
  `$ cd $GOPATH/src/github.com/kata-containers/kata-containers/tools/osbuilder/rootfs-builder`  
  `$ script -fec 'sudo -E GOPATH=$GOPATH AGENT_INIT=yes USE_DOCKER=true SECCOMP=yes ./rootfs.sh ${distro}'`

We used to use `SECCOMP=yes|no` in `rootfs.sh` to check whether `seccomp` package (`libseccomp2`) is included in the image.  
In this commit, we use it for passing the flag to the agent's `Makefile`  

## How to enoforce the seccomp

We should provide seccomp profile when we use `k8s` or `crictl` like the above [experiments](Experiments-of-the-seccomp-feature).

## Summary of the way to enable seccomp feature

| Seccomp profile provided? <sup>[1](#footnote1)</sup>| Runtime seccomp enabled? <sup>[2](#footnote2)</sup> | Agent seccomp enabled? <sup>[3](#footnote3)</sup> | Seccomp enforced? <sup>[4](#footnote4)</sup> | Notes                                                                                                                                   |
|---------------------------|--------------------------|------------------------|-------------------|-----------------------------------------------------------------------------------------------------------------------------------------|
| no                        | no                       | no                     | no                | No seccomp profile                                                                                                                      |
| no                        | yes                      | no                     | no                | No seccomp profile                                                                                                                      |
| no                        | no                       | yes                    | no                | No seccomp profile                                                                                                                      |
| no                        | yes                      | yes                    | no                | No seccomp profile                                                                                                                      |
| yes                       | no                       | no                     | no                | Seccomp profile is not passed to the runtime. Therefore, `grpc.Linux.Seccomp` is `null`.                                                  |
| yes                       | yes                      | no                     | no                | Profile provided and `grpc.Linux.seccomp` is passed to the runtime, but the agent disables seccomp feature. Therefore, an **error** occurs. |
| yes                       | no                       | yes                    | no                | Seccomp profile is not passed to the runtime.Therefore, `grpc.Linux.Seccomp` is `null`.                                                   |
| yes                       | yes                      | yes                    | yes               | Now, seccomp rule is enforced.                                                                                                                                        |

Footnotes:
- <a name="footnote1">1</a>: Implicitly or explicitly. If the seccomp profile is provided, the [seccomp object](https://github.com/opencontainers/runtime-spec/blob/master/config-linux.md#seccomp) will be present in the OCI `config.json` file.
- <a name="footnote2">2</a>: Whether the seccomp support is enabled in the runtime's `configuration.toml` configuration file.
- <a name="footnote3">3</a>: Whether the seccomp feature of the agent is enabled in the agent's `Makefile`
- <a name="footnote4">4</a>: Whether the seccomp rule will be applied.

## Notes
If we use the libseccomp crate with statically linked the libsecomp library, we should comply with [GNU LGPL-2.1](https://www.gnu.org/licenses/old-licenses/lgpl-2.1.en.html) licensed the [libseccomp library](https://github.com/seccomp/libseccomp) when we distribute the `kata agent` binary.  
Therefore, we must attach the complete source code for the libraries with annotations about the license in order to comply with the LGPL-2.1 (6(a)).

## Next actions
- Modify CI test to check this seccomp feature
- Remove the libseccomp binary (libseccomp2) in the rootfs
  - If the agent always uses the libseccomp crate with statically linked, we do not need libseccomp package in rootfs anymore.
- Update documents
- Discuss the type of linking the libseccomp library

fidencio commented:
This is really cool, I'll take a closer look at this in the coming week, @ManaSugi!

Thanks a lot for working on this one!

fidencio commented:
/test

ManaSugi commented:
Sorry.
I fixed `no newline at end of file` and removed `warning` in Makefile that I was using for my debugging.

ManaSugi commented:
/test

ManaSugi commented:
/test

mxpv commented on src/agent/rustjail/src/container.rs:
```diff
@@ -595,11 +596,20 @@ fn do_init_child(cwfd: RawFd) -> Result<()> {
+    // must have the CAP_SYS_ADMIN.
+    if linux.seccomp.is_some() && !oci_process.no_new_privileges {
+        let scmp = linux.seccomp.as_ref().unwrap();
+        seccomp::init_seccomp(scmp)?;
```
I'd suggest:
```rust
    if !oci_process.no_new_privileges {
        if let Some(ref scmp) = linux.seccomp {
            seccomp::init_seccomp(scmp)?;
        }
    }
```

ManaSugi replied:
Thanks for the review.
I fixed it.

mxpv commented on src/agent/rustjail/src/seccomp.rs:
```diff
@@ -0,0 +1,96 @@
+
+// init_seccomp creates a seccomp filter and loads it for the current process
+// including all the child processes.
+#[cfg(feature = "seccomp")]
```
You can do this at module level when importing.
```rust
#[cfg(feature = "seccomp")]
pub mod seccomp;
```

This way you can remove all feature toggles from this file and have it cleaner.

ManaSugi replied:
That makes sense.
I fixed it.

mxpv commented on src/agent/rustjail/Cargo.toml:
```diff
@@ -31,6 +31,10 @@ tokio = { version = "1.2.0", features = ["sync", "io-util", "process", "time", "
 futures = "0.3"
 async-trait = "0.1.31"
 inotify = "0.9.2"
+libseccomp = "0.1.2"
```
Let's include this only when `seccomp` feature enabled.
See examples here: https://doc.rust-lang.org/cargo/reference/features.html#optional-dependencies

ManaSugi replied:
It is cool.
Thanks for telling me.

mxpv reviewed (commented):
🎉 A couple of comments

egernst commented on src/agent/protocols/protos/agent.proto:
```diff
@@ -357,10 +357,6 @@ message AgentDetails {
-
-	// Set only if the agent is built with seccomp support and the guest
-	// environment supports seccomp.
-	bool supports_seccomp = 5;
```
Do we want to update the agent API?

ManaSugi replied:
Yes.
I think that  `supports_seccomp` flag always will be `true`, so the flag is not used anymore.
In addition, we can control whether the agent's seccomp support is enabled when building the agent.
If users want to control the flag through the agent API, we should leave it, but I'm not sure if there is such a use case.
Could you give me your thoughts?

bergwolf replied:
If we don't use a field, we should leave it here for rpc protocol compatibility. Please don't remove it.

And in fact, the field is useful on the runtime side so that we can dynamically determine if seccomp info should be passed to the agent or not. Please look at `src/runtime/virtcontainers/kata_agent.go` and see how it works currently.

ManaSugi commented on src/agent/rustjail/src/container.rs:
```diff
@@ -604,9 +605,11 @@ fn do_init_child(cwfd: RawFd) -> Result<()> {
+    #[cfg(feature = "seccomp")]
+    if !oci_process.no_new_privileges {
+        if let Some(ref scmp) = linux.seccomp {
+            seccomp::init_seccomp(scmp)?;
```
Shoud we implement the errror handling like below?

```Rust
 if !oci_process.no_new_privileges {
        if let Some(ref scmp) = linux.seccomp {
            #[cfg(feature = "seccomp")]
            seccomp::init_seccomp(scmp)?;
            #[cfg(not(feature = "seccomp"))]
            Err(anyhow!("Seccomp config provided but seccomp not supported"))  
        }
}
```
This error tells users that the runtime enables seccomp support, but the agent does not enable seccomp.

ManaSugi replied:
This error is handled in the runtime by the [commit](https://github.com/kata-containers/kata-containers/pull/1788/commits/3a41d575243631689e34c3862bc0def61ca9cad9#diff-6fb7f551035031ed29ad100c61f1a7e92b0fa15d2f67e4c7ea7ebf3b715583f4R1371).

jodh-intel replied:
I think this approach is safer: if the admin asked for seccomp but it isn't available, I'd prefer to get a hard error telling me that immediately. The current code doesn't error, which may lead an admin to _think_ seccomp is being enforced when it isn't (which would be surprising and potentially dangerous).

liubin commented:
Thank you @ManaSugi for this great PR.
I just have one question about why make `seccomp` a feature in agent source code. Only for easy delivering of agent binary for the license issue?

ManaSugi commented:
> I just have one question about why make `seccomp` a feature in agent source code. Only for easy delivering of agent binary for the license issue?
@liubin 
There are two reasons I'd like to make `seccomp` a feature in agent code.
One is making easy delivering of agent binary for the license issue as you mentioned.
Another one is that I'd like to offer users an option of disabling `seccomp` feature.
Without the option, they need to install libseccomp library (e.g. `libseccomp-dev`) on their host machine when building the agent even if they does not have plan to use `seccomp` feature.
That may be annoying for some users.
I think that some `kata` users may not need `seccomp` feature because `kata` already has strong isolation compared to other container runtimes.

liubin commented:
@ManaSugi thank you for the clarification.

liubin commented on src/agent/Makefile:
```diff
@@ -27,6 +27,14 @@ COMMIT_MSG = $(if $(COMMIT),$(COMMIT),unknown)
+##VAR SECCOMP=yes|no define if agent enables seccomp feature
+SECCOMP := no
+
+# Enable seccomp feature of rust build
```
Suggest: explicitly declare `SECCOMP_FEATURES` like `SECCOMP_FEATURES :=`

ManaSugi replied:
Thank you for the suggestion.
I fixed it and changed the name of `SECCOMP_FEATURES` to `SECCOMP_FEATURE`.

ManaSugi commented:
Now, I'm creating unit test inside the `seccomp.rs`, so I'll send it as an additional commit when I'm done.

ManaSugi commented:
This test runs only as a `root` because loading seccomp filter (`filter.load()`) fails if the user does not have `CAP_SYS_ADMIN` by unsetting `no new privileges bit`.
In addition, bump `libseccomp` crate vertion to 0.1.3.

bergwolf commented:
What is the cost of running a seccomp capable agent but do not enforce it? I'm wondering if we should just always enable the capability, and optionally enforce it when starting the agent.

ManaSugi commented:
> What is the cost of running a seccomp capable agent but do not enforce it? I'm wondering if we should just always enable the capability, and optionally enforce it when starting the agent.

@bergwolf 
Thank you for the comment.

From a performance point, the cost may be nothing.
Therefore, when we destribute the `agent` binary, it is always better to enable the seccomp feature because the users can control the feature by the runtime's `configuration.toml` when starting the kata containers.
This case is below.

| Seccomp profile provided? <sup>[1](#footnote1)</sup>| Runtime seccomp enabled? <sup>[2](#footnote2)</sup> | Agent seccomp enabled? <sup>[3](#footnote3)</sup> | Seccomp enforced? <sup>[4](#footnote4)</sup> | Notes                                                                                                                                   |
|:---------------------------|:--------------------------|:------------------------|:-------------------|:-----------------------------------------------------------------------------------------------------------------------------------------|
| yes                       | no                       | yes                    | no                | Seccomp profile is not passed to the runtime.Therefore, `grpc.Linux.Seccomp` is `null`.                                                   |

However, if we always enforce the seccomp feature in the `agent` srouce code, users who build the `agent` from the source code need to install `libsecomp` library (e.g. `libseccomp-dev`) even if they do not use the feature.
I guess it may be annoying for some users.
Therefore, I think it is better not to enforce the feature.

Could you give me your thoughts?

ManaSugi commented:
@bergwolf 
Thank you for the comment.

From a performance point, the cost may be nothing.
Therefore, when we destribute the `agent` binary, it is always better to enable the seccomp feature.                   

However, if we always enforce the seccomp feature in the `agent` srouce code, users who build the `agent` from the source code need to install `libsecomp` library (e.g. `libseccomp-dev`) even if they do not use the feature.
I guess it may be annoying for some users.
Therefore, I think it is better not to enforce the feature.

Could you give me your thoughts?

bergwolf commented:
@ManaSugi Thanks for the explanation! Then how about building it by default, and we ship the agent with seccomp capability builtin but disabled by default? And we can provide a For those who don't have libseccomp dev library and want to build agent by themselves. The benefit is that we don't have to ship two versions of kata agent (which means two versions of guest images).

ManaSugi commented:
@bergwolf 
Thanks for the replay.
Yes, the kata agent we ship is only one version that has seccomp capability.

> we ship the agent with seccomp capability builtin but disabled by default?

I have a question.
Does `disabled by default` mean `SECCOMP=no` in the agent's Makefile, doesn't it?

I'm sorry for the lack of explanation and let me check we're on the same page here.
We do not need `libseccomp-dev` in the rootfs of kata because the agent always uses the `libseccomp crate` with statically linked.
`libseccomp-dev` is required only when we build the agent (actually build `libseccomp crate`)
Therefore, we should remove the `libseccomp-dev` in the rootfs in [next actions ](https://github.com/kata-containers/kata-containers/pull/1788#issuecomment-830780832) as I mentioned eariler.
e.g.
https://github.com/kata-containers/kata-containers/blob/main/tools/osbuilder/rootfs-builder/ubuntu/config.sh#L33
I'll remove the libseccomp package inside the above script in the next pull request.

bergwolf reviewed (commented):
@ManaSugi Sorry for the delay. By `disabled by default` I mean disabling it by the runtime. The agent binary has the secomp capability built-in. The capability is exposed to the runtime and runtime can then decide if seccomp info needs to be passed to the agent. So here is my proposal:
1. we add a runtime config option to enable seccomp, and disable it by default
2. we add kata-agent build flag to control if seccomp it builtin, and enable it by default
3. for those who want seccomp support, they can just use the prebuilt agent and alter runtime config to enable it
4. for those who don't even want seccomp be built, they can build their own agent w/o seccomp support

Also please keep the `supports_seccomp` bit in the agent protocol, so that runtime can know if seccomp is supported by agent when runtime config asks for it, and fail properly if the required feature is not supported.

wdyt?

ManaSugi commented:
@bergwolf, Thank you for your kind reply and proposal.

I understand clearly and agree with your proposal. Also, I keep the `supports_seccomp` bit in the agent protocol.
I fixed the points that were pointed out.

1. I reverted the `supports_seccomp` bit.
1. I added the `seccomp` feature in the kata agent's Cargo.toml.
If the agent is built with the feature, the `rustjail`'s seccomp feature also turns on and `supports_seccomp` bit is set.
In addition, if the runtime enables seccomp in configuration.toml and the agent does not support seccomp (`supports_seccomp==false`), the runtime fails properly with an error message.
1. I enabled the seccomp feature by default.

bergwolf reviewed (approved):
lgtm! Thanks @ManaSugi !

bergwolf commented:
@ManaSugi Please rebase to solve code conflicts. Thanks!

jodh-intel commented on src/agent/Makefile:
```diff
@@ -27,6 +27,15 @@ COMMIT_MSG = $(if $(COMMIT),$(COMMIT),unknown)
+# Enable seccomp feature of rust build
+SECCOMP_FEATURE :=
+ifeq ($(SECCOMP),yes)
+    override SECCOMP_FEATURE := --features seccomp
```
Might be worth making this `EXTRA_FEATURES` to be more future-proof.

ManaSugi replied:
I changed `SECCOMP_FEATURE` to `EXTRA_RUSTFEATURES`.

jodh-intel commented on src/agent/rustjail/src/container.rs:
```diff
@@ -25,6 +25,8 @@ use crate::cgroups::mock::Manager as FsManager;
 use crate::cgroups::Manager;
 use crate::log_child;
 use crate::process::Process;
+#[cfg(feature = "seccomp")]
+use crate::seccomp;
```
If this feature is optional, we should specify whether it is available in `announce()` (for the logs). We already have a seccomp field in the `GetGuestDetails()` ttRPC response so in fact it might be worth just adding the output of `get_agent_details()` to the output of `announce()`.

jodh-intel commented on src/agent/rustjail/Cargo.toml:
```diff
@@ -31,6 +31,10 @@ tokio = { version = "1.2.0", features = ["sync", "io-util", "process", "time", "
 [dev-dependencies]
 serial_test = "0.5.0"
+
+[features]
+seccomp = ["libseccomp"]
```
This has a CI impact: strictly, we'll need a new CI with this feature enabled. Is that something we can accommodate @GabyCT and @chavafg? In fact, really we'd need one new CI _per architecture_.

ManaSugi replied:
Yes, we need to modify the CI.
Fortunately, all distributions in CI already have `libseccomp-dev` package and the agent enables seccomp capability by default.
Therefore, regarding the unit test, we just have to set the environment variables for libseccomp crate before building the agent.
e.g.
```
$ export LIBSECCOMP_LINK_TYPE=static
$ export LIBSECCOMP_LIB_PATH="the path of libseccomp.a"
```
However, when we test the seccomp feature with k8s, we have to add seccomp profile for k8s in [k8s-security-context.bats](https://github.com/kata-containers/tests/blob/main/integration/kubernetes/k8s-security-context.bats).
In that case, we also have to set `disable_guest_seccomp` as `false` in runtime's configuration like below.
`$ sudo sed -i '/^disable_guest_seccomp/ s/true/false/' /etc/kata-containers/configuration.toml`

I'm sorry if I'm wrong.

jodh-intel commented on tools/osbuilder/rootfs-builder/rootfs.sh:
```diff
@@ -579,15 +579,15 @@ EOT
 		agent_dir="${script_dir}/../../../src/agent/"
-		# For now, rust-agent doesn't support seccomp yet.
-		SECCOMP="no"
+		# The kata agent enables seccomp by default.
+		SECCOMP="yes"
```
This could look scary to some people :) I would explain in a comment that although seccomp is "enabled", it's not _enforced_ by default: you need to enable that in the main configuration file.

ManaSugi replied:
I fixed it. Thanks.

jodh-intel commented on src/agent/rustjail/src/seccomp.rs:
```diff
@@ -0,0 +1,218 @@
+
+        let data = r#"{
+            "defaultAction": "SCMP_ACT_ALLOW",
+            "architectures": [
+                "SCMP_ARCH_X86",
+                "SCMP_ARCH_X32",
+                "SCMP_ARCH_X86_64",
+                "SCMP_ARCH_AARCH64",
+                "SCMP_ARCH_ARM"
```
Kata also works on PPC64 and s390x so won't this test fail unless you add them here?

ManaSugi replied:
I added `PPC64` and `S390X` in the unit test.
To avoid endian mismatch error of `libseccomp`, I divided the architecture by endian.
Ref: https://github.com/seccomp/libseccomp/pull/138/commits/2745d8ac6b7d39051bccc8fd739808acc2741b57

jodh-intel reviewed (commented):
Thanks @ManaSugi.

A few comments.

jodh-intel commented:
Any update on this @ManaSugi?

ManaSugi commented:
I apologize for the late reply and thank you for the review @jodh-intel @bergwolf 
I updated.

ManaSugi commented:
I added a setup scirpt to install the libseccomp library and set environment variabels for the libseccomp crate.

ManaSugi commented:
I updated the above [comment](https://github.com/kata-containers/kata-containers/pull/1788#issuecomment-869179227) and the script.

ManaSugi commented:
I rebased due to the #2127.

fidencio commented:
/test

ManaSugi commented:
I created a issue on tests to pass CI.
https://github.com/kata-containers/tests/issues/3669
If there is no problem with the above issue and this PR. I'll create a commit for the issue.

jodh-intel commented on docs/Developer-Guide.md:
````diff
@@ -287,9 +309,16 @@ the following example.
+```
+$ script -fec 'sudo -E GOPATH=$GOPATH AGENT_INIT=yes USE_DOCKER=true SECCOMP=no ./rootfs.sh ${distro}'
 ```
-You MUST choose one of `alpine`, `centos`, `clearlinux`, `debian`, `euleros`, `fedora`, `suse`, and `ubuntu` for `${distro}`. By default `seccomp` packages are not included in the rootfs image. Set `SECCOMP` to `yes` to include them.
````
As shown https://github.com/kata-containers/kata-containers/tree/main/tools/osbuilder/rootfs-builder#supported-distributions-list, please could you change this to ask the user to run `$ ./rootfs.sh -l`. That avoids having to hard-code the list here (which is incorrect as it lists euleros).

ManaSugi replied:
Thank you for your comment.
I fixed it.

jodh-intel commented on docs/Developer-Guide.md:
````diff
@@ -269,6 +279,18 @@ $ go get -d -u github.com/kata-containers/kata-containers
+If you want to build the agent without the seccomp capability, you need to run `make` with `SECCOMP=no` as follows.
+
+```
+$ cd $GOPATH/src/github.com/kata-containers/kata-containers/src/agent && make SECCOMP=no
````
No problem with this, but an alternative is the following (which is slightly shorter ;):

```bash
$ make -C $GOPATH/src/github.com/kata-containers/kata-containers/src/agent SECCOMP=no
```

ManaSugi replied:
I fixed it.

jodh-intel commented on ci/install_libseccomp.sh:
```diff
@@ -0,0 +1,70 @@
+
+arch=$(uname -m)
+
+libseccomp_version="2.5.1"
+libseccomp_tarball="libseccomp-${libseccomp_version}.tar.gz"
+libseccomp_url="https://github.com/seccomp/libseccomp/releases/download/v${libseccomp_version}/${libseccomp_tarball}"
+libseccomp_install_dir="$1"
+cppflags="-O2"
+
+gperf_version="3.1"
+gperf_tarball="gperf-${gperf_version}.tar.gz"
+gperf_url="https://ftp.gnu.org/gnu/gperf/${gperf_tarball}"
+gperf_install_dir="$2"
```
Conventionally, we store external dependency details in the "versions database" [here](https://github.com/kata-containers/kata-containers/blob/main/versions.yaml). Having a central database allows us to audit it (atleast in principle ;)

We then query the versions database using the `get_version()` function here:

- https://github.com/kata-containers/tests/blob/main/.ci/lib.sh#L150

For example:

- https://github.com/kata-containers/tests/blob/main/integration/cri-o/cri-o.sh#L54

ManaSugi replied:
Thank you for the information and instruction. That makes sense.
I fixed it.

jodh-intel commented on ci/install_libseccomp.sh:
```diff
@@ -0,0 +1,70 @@
+libseccomp_tarball="libseccomp-${libseccomp_version}.tar.gz"
+libseccomp_url="https://github.com/seccomp/libseccomp/releases/download/v${libseccomp_version}/${libseccomp_tarball}"
+libseccomp_install_dir="$1"
+cppflags="-O2"
```
Unused? Should it be referenced in the hard-coded `-O2` below?

ManaSugi replied:
Sorry, `cflags` is correct. I fixed it.

jodh-intel commented on .github/workflows/static-checks.yaml:
```diff
@@ -51,6 +51,14 @@ jobs:
+      run: |
+        libseccomp_install_dir=$(mktemp -d -t libseccomp.XXXXXXXXXX)
+        gperf_install_dir=$(mktemp -d -t gperf.XXXXXXXXXX)
+        cd ${GOPATH}/src/github.com/${{ github.repository }} && ./ci/install_libseccomp.sh "${libseccomp_install_dir}" "${gperf_install_dir}"
```
The two tmp directories passed to `install_seccomp.sh` appear to be unused?!?

ManaSugi replied:
The two arguments are used at the following points.

`libseccomp_instal_dir`
https://github.com/kata-containers/kata-containers/pull/1788/files#diff-a503dc9420524f1744ca0bf10860e00742fd1b8e131ba412cc3a255a67e776b6R23

`gperf_install_dir`
https://github.com/kata-containers/kata-containers/pull/1788/files#diff-a503dc9420524f1744ca0bf10860e00742fd1b8e131ba412cc3a255a67e776b6R31

jodh-intel replied:
Ah - I'd missed that. Could you make this a bit clearer? We normally do something like this:

```bash
main()
{
  local arg="${1:-}"

  [ -z "$arg" ] && die "need foo arg"

  # ...
}

main "$@"
```

ManaSugi replied:
Sorry, I fixed it. 
Thanks.

jodh-intel commented on ci/install_libseccomp.sh:
```diff
@@ -0,0 +1,70 @@
+    build_and_install_libseccomp
+}
+
+main
```
Missing `$@` for the two temporary directories passed to the script in `static-checks.yaml`?

ManaSugi replied:
I think that it is not necessary because the arguments are referenced directly at the following points.

`libseccomp_instal_dir`
https://github.com/kata-containers/kata-containers/pull/1788/files#diff-a503dc9420524f1744ca0bf10860e00742fd1b8e131ba412cc3a255a67e776b6R23

`gperf_install_dir`
https://github.com/kata-containers/kata-containers/pull/1788/files#diff-a503dc9420524f1744ca0bf10860e00742fd1b8e131ba412cc3a255a67e776b6R31

jodh-intel commented on src/agent/rustjail/src/seccomp.rs:
```diff
@@ -0,0 +1,231 @@
+        }
+    }
+
+    // Set filtter attributes for each seccomp flag
```
Nit: typo: "filter".

ManaSugi replied:
I fixed it.

jodh-intel commented on src/agent/rustjail/src/seccomp.rs:
```diff
@@ -0,0 +1,231 @@
+    filter.set_no_new_privs_bit(false)?;
+
+    // Add a rule for each system call
+    for syscall in &scmp.syscalls {
+        if syscall.names.is_empty() {
+            return Err(anyhow!("syscall name is required"));
+        }
+
+        let action = ScmpAction::from_str(&syscall.action, Some(syscall.errno_ret))?;
+        if action == def_action {
+            continue;
+        }
+
+        for name in &syscall.names {
+            let syscall_num = get_syscall_from_name(name, None)?;
+
+            if syscall.args.is_empty() {
+                filter.add_rule(action, syscall_num, None)?;
+            } else {
+                let mut cmps: Vec<ScmpArgCompare> = Vec::new();
+
+                for arg in &syscall.args {
```
This code is potentially too complex as we have three levels of nested loops. Is there any way you could refactor to call a `handle_syscall()` function or something like that?

ManaSugi replied:
That’s true. I added `get_rule_conditions` funcition to avoid the three levels of nested loops.

jodh-intel reviewed (commented):
Thanks @ManaSugi. A few more comments.

ManaSugi commented:
Thank you for your review. @jodh-intel 
I applied the review feedback changes.
Besides, I rebased for #2197 and #2203.

fidencio commented:
/test

jodh-intel commented on src/runtime/virtcontainers/kata_agent.go:
```diff
@@ -1420,6 +1420,10 @@ func (k *kataAgent) createContainer(ctx context.Context, sandbox *Sandbox, c *Co
 	sharedPidNs := k.handlePidNamespace(grpcSpec, sandbox)
 
+	if !sandbox.config.DisableGuestSeccomp && !sandbox.seccompSupported {
+		return nil, fmt.Errorf("Seccomp profiles are passed to the virtual machine, but the kata agent does not support seccomp")
```
Nit: "kata" -> "Kata".

ManaSugi replied:
I fixed it. Thanks

jodh-intel reviewed (approved):
Thanks @ManaSugi!

lgtm

ping @kata-containers/kata-containers for more reviews.

ManaSugi commented:
I'd like to confirm one thing.
Currently, the kata-agent is built on the host, not the docker container as mentioned [here](https://github.com/kata-containers/kata-containers/pull/2171#issuecomment-872968124).
Hence, the `libeseccomp` library (and `gperf`) is also built and installed in temporal directories that are removed after the agent build is complete on the host. However, If the agent build on the docker is re-enabled, the libseccomp should be also built on it.
Do you have plans for re-enabling it? If not, I'd like to work on [the issue](https://github.com/kata-containers/tests/issues/3669) in order to pass CI.

Jakob-Naucke commented on src/agent/rustjail/src/seccomp.rs:
```diff
@@ -0,0 +1,237 @@
+        let mut scmp: oci::LinuxSeccomp = serde_json::from_str(data).unwrap();
+        let mut arch: Vec<oci::Arch>;
+
+        if cfg!(target_endian = "little") {
```
Nice catch @ManaSugi! I've tested this on s390x, it's working, and the tests are also passing.

ManaSugi replied:
Thank you for your approval and testing this on s390x @Jakob-Naucke!

liubin commented:
Hi @ManaSugi @fidencio @Jakob-Naucke  so what's the conclusion?

The situation for now:

- #2171 is a cleanup, it deletes golang related things for building the agent.
- Now the agent is built on the host side, and it **should** be built in the container.
- This PR installs new packages on the host too.

So next, we will

- Let #2171 be got merged
- Let this PR be got merged
- Raise a new PR to re-enable building agent in the container and install packages for `libeseccomp` 

Don't know if my understanding is correct, please share your opinions.

Jakob-Naucke commented:
> So next, we will
> - Let [osbuilder: Skip installing golang for building rootfs #2171](https://github.com/kata-containers/kata-containers/pull/2171) be got merged
> - Let this PR be got merged
> - Raise a new PR to re-enable building agent in the container and install packages for `libeseccomp`


@liubin yes, this was my understanding.

ManaSugi commented:
Thank you for your comment @liubin.
I agree with your opinion too.

ManaSugi commented:
I squashed commits by subsystem.

ManaSugi commented:
I squashed the commits by subsystem.

jodh-intel commented:
/test

cmaf commented:
@ManaSugi could you please rebase to resolve code conflicts?

ManaSugi commented:
[#3724](https://github.com/kata-containers/tests/pull/3724) was merged, so I rebased.

ManaSugi commented:
/test

jodh-intel commented:
Ok - this PR is now mergeable but we still have four commenters who haven't approved, so please could one of @lifupan, @liubin , @mxpv and @egernst take a final look and ack before we hit the merge button?

Jakob-Naucke commented:
@ManaSugi I'm afraid yes, we have this problem in a bunch of instances
